File: lox.net/lox.net/Token.cs
namespace lox.net
{
    public class Token
    {
        public Token(
            TokenType type,
            string lexeme,
            object? literal,
            int line)
        {
            Type = type;
            Lexeme = lexeme;
            Literal = literal;
            Line = line;
        }

        public TokenType Type { get; }
        public string Lexeme { get; }
        public object? Literal { get; }
        public int Line { get; }

        public override bool Equals(object? obj)
        {
            var other = obj as Token;
            return other != null && this.Type == other.Type && this.Lexeme == other.Lexeme;
        }

        public override int GetHashCode() => Type.GetHashCode() * 31 + Lexeme.GetHashCode();

        public override string ToString()
        {
            return $"{Type} {Lexeme} {Literal}";
        }
    }
}

File: kotlin/klox/src/main/kotlin/eu/grigoriu/craftinginterpreters/klox/Token.kt
package eu.grigoriu.craftinginterpreters.klox

class Token(
    val type: TokenType,
    val lexeme: String,
    // Can we make this generic?
    val literal: Any?,
    val line: Int
) {
    override fun toString(): String {
        return "$type $lexeme $literal"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Token

        if (type != other.type) return false
        if (lexeme != other.lexeme) return false
        if (literal != other.literal) return false

        return true
    }

    override fun hashCode(): Int {
        var result = type.hashCode()
        result = 31 * result + lexeme.hashCode()
        result = 31 * result + (literal?.hashCode() ?: 0)
        return result
    }
}